src/tree.rs

```diff
@@ -135,3 +138,159 @@ pub fn minimum_spanning_tree(
+                continue;
+            }
+            let pop = pops[node.index()];
+
+            // todo: factor out expensive clones
+            // Mark as seen; push to queue if only one unseen neighbor
+            let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+                .graph
+                .neighbors(node)
+                .filter(|node| !seen_nodes[node.index()])
+                .collect();
+            // eprintln!("unseen_neighbors: {}", unseen_neighbors.len());
+            if unseen_neighbors.len() == 1 {
+                // this will be false if root
+                let neighbor = unseen_neighbors[0];
+                pops[neighbor.index()] += pop.clone();
+                let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
+                same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+                // eprintln!("node pushed to queue (pop = {}, target = {}): {}", pops[neighbor.index()], pop_target, neighbor.index());
+
+                if !node_queue.contains(&neighbor) {
+                    node_queue.push_back(neighbor);
+                }
+            } else if unseen_neighbors.len() == 0 {
+                break;
+            } else {
+                continue;
+            }
+            // pops[node.index()] = 0.0; // not needed?
+
+            // Check if balanced
+            if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+                // slightly different
+                // eprintln!("balanced node found: {}", node.index());
+                balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
+            }
+
+            seen_nodes[node.index()] = true;
+        }
+        balanced_nodes
+    });
+
+    Ok(balanced_nodes)
+}
+
+/// Bipartition graph into two contiguous, population-balanced components.
+/// Assumes that graph is contiguous.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. This
+///     tells retworkx/rust how to extract a numerical weight as a ``float``
+///     for edge object. Some simple examples are::
+///
+///         minimum_spanning_tree(graph, weight_fn: lambda x: 1)
+///
+///     to return a weight of 1 for all edges. Also::
+///
+///         minimum_spanning_tree(graph, weight_fn: float)
+///
+///     to cast the edge object as a float as the weight.
+/// :param float default_weight: If ``weight_fn`` isn't specified this optional
+///     float value will be used for the weight/cost of each edge.
+///
+/// :returns: A set of nodes in one half of the spanning tree
+///
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, pop, target_pop, epsilon)")]
+pub fn bipartition_tree(
+    py: Python,
+    graph: &graph::PyGraph,
+    weight_fn: PyObject,
+    py_pops: &PyList,
```

Comment:
If this is readonly I would suggest using a numpy array input type here:

```suggestion
    py_pops: PyReadonlyArray1<f64> ,
```

Then we can convert this to a slice with `let pops_slice = py_pops.as_slice()?;`, they we can pass `&[f64]` to the inner `balanced_cut_edge()` call and everything is passed by reference from the inner c array in numpy so we have very little overhead here. Using a list (or a vec as in my other comment) you have iteration and copy/conversion overhead going from a list of floats in python to a Vec<f64> in rust.

Reply:
Hmm, I'm getting this error when I try that: 
```
the trait `PyClass` is not implemented for `&[f64]`
```
is there something else I should be passing?

Reply:
oh that was my bad on the earlier comment I missed that `balance_cut_edge()` was also a pyfunction I don't think you can have `&[f64]` as an input type to a pyfunction because the pyo3 macro can convert to that directly. We can just pass the `PyReadOnlyArray1` directly to `balance_cut_edge()` and then do the slice cast call in `balance_cut_edge()`

Reply:
Ok. I just realized that this isn't read-only because I re-use the same vector to accumulate populations during each loop (when I do node contraction). Do you have any suggestions on minimizing the number of copies/memory allocs made per loop in this case?

Reply:
To clarify: I don't seem to be able to pass mutable vector references to pyfunctions, so the current code copies the vector over twice (I think). E.g. I want to get rid of this: https://github.com/Qiskit/retworkx/pull/572/files#diff-9794019b962b882943346745baf52516c9f655c4198f992f64bfdb6bd545f9d2R172

Reply:
Ah, figured it out.